lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs

```diff
@@ -175,8 +176,57 @@ liftHandler action = Handler (withExceptT toServerError action)
 liftE :: IsServerError e => e -> Handler a
 liftE = liftHandler . throwE
 
-apiError :: ServerError -> ApiErrorInfo -> Text -> ServerError
-apiError err info messageUnformatted = err
+-- | Current, blessed way of creating API errors, with rich error info objects.
+--
+-- Note that errors constructed with this function do /not/ include an error
+-- message, by design.
+--
+-- Instead, they include an optional error info object.
+--
+-- All information necessary to interpret every error should be available from:
+--
+--   - the swagger specification (which should include both a schema and
+--     human-readable description of the error).
+--
+--   - the serialised name of the error (which enables the caller to look up
+--     the error in the swagger specification).
+--
+--   - the optional rich error info object.
+--
+-- Usage example:
+--
+-- @
+-- apiError err403 $ UtxoTooSmall
+--     ApiErrorTxOutputLovelaceInsufficient
+--         { txOutputIndex = index
+--         , txOutputLovelaceSpecified =
+--             Coin.toQuantity $ TokenBundle.getCoin $ snd $ view #output e
+--         , txOutputLovelaceRequiredMinimum =
+--             Coin.toQuantity $ view #minimumExpectedCoin e
+--         }
+-- @
+--
+apiError :: ServerError -> ApiErrorInfo -> ServerError
+apiError err info = err
+    { errBody = Aeson.encode ApiError {info, message}
+    , errHeaders =
+        (hContentType, renderHeader $ contentType $ Proxy @JSON)
+        : errHeaders err
+    }
+  where
+    -- For backwards compatibility with our JSON roundtrip test suite, we have
+    -- to include a message string. We can remove this requirement later on,
+    -- once all errors have been converted to use 'apiError' instead of
+    -- 'apiErrorOldDeprecated'.
+    message = ApiErrorMessage ""
```

Comment:
Don't we want messages though as some form of human readable description?

For instance, the `ApiErrorBalanceTxUnderestimatedFee` `info` object I just added ended up pretty verbose:

```haskell
data ApiErrorBalanceTxUnderestimatedFee = ApiErrorBalanceTxUnderestimatedFee
    { underestimation :: !(Quantity "lovelace" Natural)
    , estimatedNumberOfKeyWits :: Natural
    , estimatedNumberOfBootstrapKeyWits :: Natural
    , candidateTxHex :: Text
    , candidateTxReadable :: Text
    }
    deriving (Data, Eq, Generic, Show, Typeable)
    deriving (FromJSON, ToJSON)
        via DefaultRecord ApiErrorBalanceTxUnderestimatedFee
    deriving anyclass NFData
```

So I kept the following as `message`
```
I have somehow underestimated the fee of the transaction by 0.019000 and cannot finish balancing
```
which I thought was a nice summary.

Reply:
Because the message should be a function of the rich object, I wonder if we could move up to some type level. But in general good point.

Reply:
> Don't we want messages though as some form of human readable description?

Hi @Anviking, thanks for commenting!

I do agree that it's valuable to have a human-readable description of the error.

My main objection is that we define human-readable descriptions in **_multiple_** places. For example, for the `output_token_quantity_exceeds_limit` HTTP API error, we define at least **_three_** different human-readable descriptions:

1. One in `swagger.yaml`:
    https://github.com/cardano-foundation/cardano-wallet/blob/af1b3b715048b3282e063097786c11490c40c306/specifications/api/swagger.yaml#L4704-L4713
2. A second in `Http/Server/Error.hs`:
    https://github.com/cardano-foundation/cardano-wallet/blob/af1b3b715048b3282e063097786c11490c40c306/lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs#L1013-L1031
3. And a third in `Framework/TestData.hs`:
    https://github.com/cardano-foundation/cardano-wallet/blob/af1b3b715048b3282e063097786c11490c40c306/lib/integration/framework/Test/Integration/Framework/TestData.hs#L567-L583

These messages all say more or less the same thing: that we have a quantity of some asset that exceeds the upper limit of what can be included in a single output.

Instead of maintaining these multiple descriptions in different places (giving us multiple sources of truth), I'd propose that:

- we define just a single human-readable description for each error;
- that we locate each human-readable description within the OpenAPI specification.

If we also want to **_return_** a human-readable message **_with_** each error, then perhaps we can automatically return the value of the corresponding `description` field from `swagger.yaml` along with each error object?

What do you think?

Reply:
> My main objection is that we define these human-readable descriptions in multiple places. For example, for the output_token_quantity_exceeds_limit HTTP API error, we define at least three different human-readable descriptions:

I don't disagree 👍 

> that we locate each human-readable description within the OpenAPI specification.

That might make sense... will the reading of the OpenAPI specification from the wallet server implementation turn out nicely though?

Reply:
> will the reading of the OpenAPI specification from the wallet server implementation turn out nicely though?

@Anviking I'm not sure, as I haven't tried this yet!

But in principle, I imagine we could use the [openapi3](https://hackage.haskell.org/package/openapi3) package (or something similar) to look up any error and find the value of its `description`. This `description` could be returned in cases where the `message` is missing or empty.

I would propose to tackle that as a separate PR. (I'm happy to do it, but probably as a background task!)